### PR TITLE
Remove jsonapi and replace it with serde

### LINKED ISSUE #69: Replace jsonapi crate with plain serde_json
I started off using the jsonapi crate to get going, but given that we don't use all of the spec, it's a little complex for our needs. In addition, it uses a lot of dynamic dispatch, hashmaps, etc that we wouldn't need.

Fundamentally, the JSON looks like this:

```json
{
  "data": {
    "type": "crate",
    "id": "example",
    "attributes": {
      "docs": " docs go here"
    },
    "relationships": {
      "modules": {
        "data": [
          {
            "type": "module",
            "id": "example::example_module"
          }
        ]
      }
    }
  },
  "included": [
    {
      "type": "module",
      "id": "example::example_module",
      "attributes": {
        "docs": " docs go here",
        "name": "example_module"
      }
    }
  ]
}
```

Let's break this down a bit:

at the top level, we have two keys: `data` and `included`. `included` is an array of documents, `data` is a single one. `data` is for the crate itself, `included` is for everything else.

Each document looks like this:

```json
{
    "type": "crate",
    "id": "example",
    "attributes": {
      "docs": " docs go here"
    },
    "relationships": {
      "modules": {
        "data": [
          {
            "type": "module",
            "id": "example::example_module"
          }
        ]
      }
    }
}
```

There are four keys:

1. "type" contains the type, so right now, the top level is "crate" and everything else is "module". Eventually this will expand to the other items, but we don't do that just yet.
2. "id" is called the "qualified name" in rls-analysis. A globally unique identifier. More about this in "relationships" below.
3. "attributes" contains all of the actual data for the thing. Right now, we only have two attributes: "docs", which is the doc string, and "name", which is just the name, without the parent stuff. compare the name to the id of the included module above, for example.
4. "relationships" establishes the links between stuff. In the first example, there are no sub-crates, and so it has no `relationships`, the only one is in the crate itself. If `example_module` had a sub-module, it'd have a relationship to it; basically, the stuff without relationships is the leafs of the tree. You can also think of this as a foreign key relationship in a database.

Some more on relationships, since they have their own structure:

```json
      "modules": {
        "data": [
          {
            "type": "module",
            "id": "example::example_module"
          }
        ]
      }
```

At the top levels are keys for the types of relationships that they have; in this case, we only have modules, so there's only one, but that expands in the future.

Inside that is one more key, "data", which is is an array of all of the related resources, one JSON object per link. Those objects have a "type" and an "id". This should match with one thing inside of `included`, and the combination of the two provide the link. 

---------------------

For super details, http://jsonapi.org/ is the spec; it contains everything but the `attributes`, which are free-form and are what we're using to display the actual docs. http://jsonapi.org/faq/#is-there-a-json-schema-describing-json-api has a JSON-SCHEMA that can be used to verify that things match up and are reasonable too.

@mgattozzi said he'd be interested in implementing this. I'm happy to elaborate if the above isn't clear in some way!

### PR DESCRIPTION
**Pull request by mgattozzi**
This changes rustdoc over completely to serde. While jsonapi was a great
crate to get started it had many features and things we weren't going to
use at all. Instead we implemented a limited subset of the spec. This
moves us over to just using serde to output data about the crate. A few
things worth noting:

- The `Crate` type was merged into the `DocData` type since it was only
  used in that type. This made the resulting serialization code much
  cleaner by removing references to krate.
- The builder method was used as a way to cleanly build up the types
  for eventual serialization and because it's readable.
- Using `&str` vs String was chosen to avoid allocations that weren't
  needed
- `HashMaps` were made with the capacity of given constants since we
  know how many items at most will exist there.

Overall this commit makes serializing data easier to work with and
more reabable then before since we don't need to use all of jsonapi's
more advanced features. The nice thing is we can still use jsonapi to
validate the output that we create!

Closes #69 (Nice)

@steveklabnik Unfortunately this is a bit bigger of a PR and harder to review due to the nature of having to rip things out and replacing them. Hopefully it's not too bad but please leave any feedback and I'll be happy to update this PR.

### CODE REVIEW

File: src/json.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Document {
+    #[serde(rename = "type")]
+    type_: String,
+    id: String,
+    attributes: HashMap<String, String>,
+    relationships: Option<HashMap<String, HashMap<String, Vec<Data>>>>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
```
Comment (mgattozzi) [inline src/json.rs]: I can actually get rid of `Clone` here. It's not needed anymore.

**Comment by DevOrc** [comment]
This will probably conflict with #83 because it uses the jsonapi to validate it... So you could probably just add it as a dev-dependency? I don't know if there is a equivalent of the [validate function](https://docs.rs/jsonapi/0.5.2/jsonapi/api/struct.JsonApiDocument.html#method.validate) is serde...

**Comment by mgattozzi** [comment]
I put it in as a dev dependency and it worked perfectly fine as is locally :D No need to worry about it now.

**Comment by steveklabnik** [comment]
Yup!

@mgattozzi I'm going to review this shortly, it's the biggest PR of the weekend ones so it was easier to knock those out!

**Comment by mgattozzi** [comment]
@steveklabnik Not a problem! I just rebased and cleaned everything up and pushed it with all the changes I made rolled into the one commit for you to review.

@DevOrc thanks for the json validation test. Definitely helps knowing that what I cooked up actually works correctly :D

File: src/json.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
+//! Code used to serialize crate data to JSON. We use a subset of the JSON-API spec.
+//!
```
Comment (steveklabnik) [inline src/json.rs]: ✂️

File: src/json.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+#[derive(Serialize, Debug)]
+pub struct Documentation<'a> {
+    data: Option<Document<'a>>,
+    included: Option<Vec<Document<'a>>>,
+}
+
+#[derive(Serialize, Debug)]
+pub struct Document<'a> {
+    #[serde(rename = "type")]
+    type_: &'a str,
```
Comment (steveklabnik) [inline src/json.rs]: i'd prefer `ty`

**Comment by steveklabnik** [comment]
:tada: thanks so much!

**Comment by mgattozzi** [comment]
You're welcome :D

**Pull request merged**